### PR TITLE
Dbg diag

### LINKED ISSUE #513: occassional advance test failure
```
  ======================================================================
10:05:21
  FAIL: test_overlapped_particledatas_have_identical_particles_1 (__main__.AdvanceTest)
10:05:21
  ----------------------------------------------------------------------
10:05:21
  Traceback (most recent call last):
10:05:21
    File "/usr/lib/python3.8/site-packages/ddt.py", line 145, in wrapper
10:05:21
      return func(self, *args, **kwargs)
10:05:21
    File "test_advance.py", line 420, in test_overlapped_particledatas_have_identical_particles
10:05:21
      self._test_overlapped_particledatas_have_identical_particles(dim, interp_order, refinement_boxes)
10:05:21
    File "test_advance.py", line 409, in _test_overlapped_particledatas_have_identical_particles
10:05:21
      self.assertTrue(np.allclose(part1.v[idx1,0], part2.v[idx2,0], atol=1e-12))
10:05:21
  AssertionError: False is not true
10:05:21
  
10:05:21
  ----------------------------------------------------------------------
10:05:21
  Ran 17 tests in 47.217s
10:05:21
  
10:05:21
  FAILED (failures=1)

```
```

 ======================================================================
09:08:44
  FAIL: test_overlapped_particledatas_have_identical_particles_2 (__main__.AdvanceTest)
09:08:44
  ----------------------------------------------------------------------
09:08:44
  Traceback (most recent call last):
09:08:44
    File "/usr/lib/python3.8/site-packages/ddt.py", line 145, in wrapper
09:08:44
      return func(self, *args, **kwargs)
09:08:44
    File "test_advance.py", line 420, in test_overlapped_particledatas_have_identical_particles
09:08:44
      self._test_overlapped_particledatas_have_identical_particles(dim, interp_order, refinement_boxes)
09:08:44
    File "test_advance.py", line 409, in _test_overlapped_particledatas_have_identical_particles
09:08:44
      self.assertTrue(np.allclose(part1.v[idx1,0], part2.v[idx2,0], atol=1e-12))
09:08:44
  AssertionError: False is not true
```

### PR DESCRIPTION
**Pull request by nicolasaunai**
this PR intends to fix  #513 

where phare_overlapped_particledatas_have_identical_particles_1 fails because two particles are exactly at the same position.

- changes particle delta from float to double.
- dump diagnostics in double precision when in debug mode or if `-DCMAKE_CXX_FLAGS="-DPHARE_DIAG_DOUBLES=1"`  is passed to cmake

when compiled with double precision dumps and with the edits of this PR the following command runs without failure

``` 
for I in $(seq 1 1000); do echo $I; python3 -u tests/simulator/test_advance.py AdvanceTest.test_overlapped_particledatas_have_identical_particles_1   || break; done
``` 

after the edits in this PR but having float dumps leads to a 1-2% failure rate

The conclusion is therefore that float precision dump and float delta has a non-negligible probability to get two particles at the same position. Note this occurs in 1D, should be less of a problem in 2 and 3D.

### CODE REVIEW

File: pyphare/pyphare/pharesee/particles.py [hunk 1]
```diff
@@ -122,6 +140,33 @@ def isin(p, b):
                          charges=self.charges[idx],
                          dl = self.dl[idx])
 
+    def __getitem__(self, box):
+        return self.select(box)
+
+
+    def erase(self, idx_or_list):
+        idx_list = [idx_or_list] if is_scalar(idx_or_list) else idx_or_list
```
Comment (PhilipDeegan) [inline pyphare/pyphare/pharesee/particles.py:148]: ```def erase(self, idx_list):```

the if is_scalar check is probably not needed

File: pyphare/pyphare/pharesee/particles.py [hunk 2]
```diff
@@ -122,6 +140,33 @@ def isin(p, b):
...
+        self.iCells  = np.delete(self.iCells, idx_list, axis=0)
+        self.deltas  = np.delete(self.deltas, idx_list, axis=0)
+        self.v       = np.delete(self.v, idx_list, axis=0)
+        self.weights = np.delete(self.weights, idx_list, axis=0)
+        self.charges = np.delete(self.charges, idx_list, axis=0)
+        self.dl      = np.delete(self.dl, idx_list, axis=0)
+
+
+    def pop(self, idx_or_list):
```
Comment (PhilipDeegan) [inline pyphare/pyphare/pharesee/particles.py:157]: same here

File: src/core/data/particles/particle_utilities.h [hunk 3]
```diff
@@ -33,6 +33,60 @@ namespace core
         }
```
Comment (PhilipDeegan) [inline src/core/data/particles/particle_utilities.h:33]: might aswell change this file to "namespace PHARE::core {"

while you're at it?

File: src/core/numerics/pusher/boris.h [hunk 4]
```diff
@@ -5,10 +5,14 @@
 #include <cmath>
 #include <cstddef>
 #include <iostream>
-
+#include <iomanip>
+#include <algorithm>
+#include <iterator>
+#include <stdexcept>
```
Comment (PhilipDeegan) [inline src/core/numerics/pusher/boris.h:11]: still needed?

File: tests/simulator/test_advance.py [hunk 5]
```diff
@@ -63,10 +62,10 @@ def by(x):
             L = global_vars.sim.simulation_domain()[0]
             v1=-1
             v2=1.
-            return v1 + (v2-v1)*(S(x,L*0.25,1) -S(x, L*0.75, 1))
+            return 0.#v1 + (v2-v1)*(S(x,L*0.25,1) -S(x, L*0.75, 1))
```
Comment (PhilipDeegan) [inline tests/simulator/test_advance.py:65]: rm comment, or replace 0 with commented out part?

File: pyphare/pyphare/pharesee/particles.py [hunk 6]
```diff
@@ -122,6 +140,31 @@ def isin(p, b):
...
+    def erase(self, idx):
+        self.iCells  = np.delete(self.iCells,idx, axis=0)
+        self.deltas  = np.delete(self.deltas,idx, axis=0)
+        self.v       = np.delete(self.v,idx, axis=0)
+        self.weights = np.delete(self.weights,idx, axis=0)
+        self.charges = np.delete(self.charges,idx, axis=0)
+        self.dl      = np.delete(self.dl,idx, axis=0)
+
+
+    def pop(self, idx_or_list):
```
Comment (PhilipDeegan) [inline pyphare/pyphare/pharesee/particles.py:156]: :upside_down_face:

Comment (nicolasaunai) [reply pyphare/pyphare/pharesee/particles.py]: yes, changed thx

**Review by PhilipDeegan — APPROVED** [review]

**Pull request merged**